model_compression_toolkit/core/common/mixed_precision/solution_refinement_procedure.py

```diff
@@ -45,26 +58,66 @@ def greedy_solution_refinement_procedure(mp_solution: List[int],
     while changed:
         changed = False
         nodes_kpis = {}
+        nodes_next_candidate = {}
+
         for node_idx in range(len(mp_solution)):
             if new_solution[node_idx] == 0:
                 # layer has max config in the given solution, nothing to optimize
                 continue
 
-            node_next_candidate_idx = new_solution[node_idx] - 1
-            node_updated_kpis = search_manager.compute_kpi_for_config(
-                config=search_manager.replace_config_in_index(new_solution, node_idx, node_next_candidate_idx))
+            node_candidates = copy.deepcopy(search_manager.graph.get_configurable_sorted_nodes()[node_idx].candidates_quantization_cfg)
+            valid_candidates = _get_valid_candidates_indices(node_candidates, new_solution[node_idx])
+
+            # Create a list of KPIs for the valid candidates.
+            valid_kpis = []
+            for valid_idx in valid_candidates:
+                node_updated_kpis = search_manager.compute_kpi_for_config(
+                    config=search_manager.replace_config_in_index(new_solution, node_idx, valid_idx))
+                valid_kpis.append(node_updated_kpis)
+
+            # filter out new configs that don't hold the KPI restrictions
+            nodes_kpis_list = [(node_idx, kpis) for node_idx, kpis in zip(valid_candidates,valid_kpis) if
```

Comment:
The name "nodes_kpis_list" is confusing because there is a different variable with that name in the outer scope.
Consider changing to "node_filtered_kpis" (node singular)